pkg/api/pluginproxy/access_token_provider_test.go

```diff
@@ -91,4 +95,192 @@ func TestAccessToken(t *testing.T) {
 			So(token2, ShouldEqual, "abc")
 		})
 	})
+
+	Convey("Plugin with token auth route", t, func() {
+		apiHandler := http.NewServeMux()
+		server := httptest.NewServer(apiHandler)
+
+		defer server.Close()
+		pluginRoute := &plugins.AppPluginRoute{
+			Path:   "pathwithtokenauth1",
+			Url:    "",
+			Method: "GET",
+			TokenAuth: &plugins.JwtTokenAuth{
+				Url: server.URL + "/oauth/token",
+				Scopes: []string{
+					"https://www.testapi.com/auth/monitoring.read",
+					"https://www.testapi.com/auth/cloudplatformprojects.readonly",
+				},
+				Params: map[string]string{
+					"grant_type":    "client_credentials",
+					"client_id":     "{{.JsonData.client_id}}",
+					"client_secret": "{{.SecureJsonData.client_secret}}",
+					"audience":      "{{.JsonData.audience}}",
+					"client_name":   "datasource_plugin",
+				},
+			},
+		}
+
+		templateData := templateData{
+			JsonData: map[string]interface{}{
+				"client_id": "my_client_id",
+				"audience":  "www.example.com",
+			},
+			SecureJsonData: map[string]string{
+				"client_secret": "my_secret",
+			},
+		}
+
+		ds := &models.DataSource{Id: 1, Version: 2}
+
+		var token map[string]interface{}
+
+		var i int
+		apiHandler.HandleFunc("/oauth/token", func(w http.ResponseWriter, req *http.Request) {
+			json.NewEncoder(w).Encode(token)
+			i++
+		})
+
+		Convey("Should cache token", func() {
+			Convey("Should parse and cache access token with expires_in string field", func() {
+				token = map[string]interface{}{
+					"access_token":  "2YotnFZFEjr1zCsicMWpAA",
+					"token_type":    "example",
+					"expires_in":    "3600",
+					"refresh_token": "tGzv3JOkF0XG5Qx2TlKWIA",
+				}
+
+				provider := newAccessTokenProvider(ds, pluginRoute)
+				t := time.Now()
+				accessToken, err := provider.getAccessToken(context.Background(), templateData)
+				So(err, ShouldBeNil)
+				So(accessToken, ShouldEqual, token["access_token"])
+
+				accessToken, err = provider.getAccessToken(context.Background(), templateData)
+				So(err, ShouldBeNil)
+				So(accessToken, ShouldEqual, token["access_token"])
+
+				for _, v := range tokenCache.cache {
+					So(v.ExpiresOn.Unix(), ShouldAlmostEqual, t.Unix()+3600)
+				}
```

Comment:
Instead of looping over all cached tokens, you would better look for the specific token using the provided key.